include/podio/DataSource.h

```diff
@@ -0,0 +1,179 @@
+#include <ROOT/RDataFrame.hxx>
+#include <ROOT/RDataSource.hxx>
+
+// STL
+#include <memory>
+#include <string>
+#include <typeinfo>
+#include <utility>
+#include <vector>
+
+namespace podio {
+class DataSource : public ROOT::RDF::RDataSource {
+public:
+  ///
+  /// @brief Construct the podio::DataSource from the provided file.
+  ///
+  explicit DataSource(const std::string& filePath, int nEvents = -1);
+
+  ///
+  /// @brief Construct the podio::DataSource from the provided file list.
+  ///
+  explicit DataSource(const std::vector<std::string>& filePathList, int nEvents = -1);
+
+  ///
+  /// @brief Inform the podio::DataSource of the desired level of parallelism.
+  ///
+  void SetNSlots(unsigned int nSlots) override;
+
+  ///
+  /// @brief Retrieve from podio::DataSource per-thread readers for the desired
+  /// columns.
+  ///
+  template <typename T>
+  std::vector<T**> GetColumnReaders(std::string_view columnName);
+
+  ///
+  /// @brief Inform podio::DataSource that an event-loop is about to start.
+  ///
+  void Initialize() override;
+
+  ///
+  /// @brief Retrieve from podio::DataSource a set of ranges of entries that
+  ///        can be processed concurrently.
+  ///
+  std::vector<std::pair<ULong64_t, ULong64_t>> GetEntryRanges() override;
+
+  ///
+  /// @brief Inform podio::DataSource that a certain thread is about to start
+  ///        working on a certain range of entries.
+  ///
+  void InitSlot(unsigned int slot, ULong64_t firstEntry) override;
+
+  ///
+  /// @brief Inform podio::DataSource that a certain thread is about to start
+  ///        working on a certain entry.
+  ///
+  bool SetEntry(unsigned int slot, ULong64_t entry) override;
+
+  ///
+  /// @brief Inform podio::DataSource that a certain thread finished working
+  ///        on a certain range of entries.
+  ///
+  void FinalizeSlot(unsigned int slot) override;
+
+  ///
+  /// @brief Inform podio::DataSource that an event-loop finished.
+  ///
+  void Finalize() override;
+
+  ///
+  /// @brief Returns a reference to the collection of the dataset's column
+  ///        names
+  ///
+  const std::vector<std::string>& GetColumnNames() const override;
+
+  ///
+  /// @brief Checks if the dataset has a certain column.
+  ///
+  bool HasColumn(std::string_view columnName) const override;
+
+  ///
+  /// @brief Type of a column as a string. Required for JITting.
+  ///
+  std::string GetTypeName(std::string_view columnName) const override;
+
+protected:
+  ///
+  /// @brief Type-erased vector of pointers to pointers to column
+  ///        values --- one per slot.
+  ///
+  std::vector<void*> GetColumnReadersImpl(std::string_view name, const std::type_info& typeInfo) override;
+
+  std::string AsString() override {
+    return "Podio data source";
+  };
+
+private:
+  /// Number of slots/threads
+  unsigned int m_nSlots = 1;
+
+  /// Input filename
+  std::vector<std::string> m_filePathList = {};
+
+  /// Total number of events
+  unsigned int m_nEvents = 0;
```

Comment:
```suggestion
  ULong64_t m_nEvents = 0;
```
Shouldn't this match with the ranges defined in the next lines?

Reply:
Changed, thx.